### PR TITLE
feat: Add saber eject-theme command

### LINKED ISSUE #247: Add theme eject command
[Issuehunt badges](https://issuehunt.io/r/saberland/saber/issues/247)

## Feature request

Add `saber eject`, similar to `vuepress eject` to copy over the currently used theme from `node_modules` or from Github (perhaps as a submodule) to `theme` and change the config file accordingly.

#### What problem does this feature solve?

Allows modification of installed themes.

#### What does the proposed API look like?

`saber eject`

#### How should this be implemented in your opinion?

Give the user the choice whether to use the perhaps precompiled files from `node_modules` or to clone the theme's repo as a submodule.

#### Are you willing to work on this yourself?

Yes.

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Sponsors (Total: $105.00)

- [egoist egoist](https://issuehunt.io/u/egoist) ($5.00)
- [issuehunt issuehunt](https://issuehunt.io/u/issuehunt) ($100.00)

### Tips

- Checkout the [Issuehunt explorer](https://issuehunt.io/r/saberland/saber/) to discover more funded issues.
- Need some help from other developers? [Add your repositories](https://issuehunt.io/r/new) on IssueHunt to raise funds.
</details>

### PR DESCRIPTION
**Pull request by krmax44**
**Summary**

Closes #247 

**What kind of change does this PR introduce?** (check at least one)

- [ ] Bugfix
- [x] Feature
- [ ] Code style update
- [ ] Refactor
- [x] Docs
- [ ] Build-related changes
- [ ] Other, please describe:

If changing the UI/CSS related code, please provide the **before/after** screenshot:

**Does this PR introduce a breaking change?** (check one)

- [ ] Yes
- [x] No

If yes, please describe the impact and migration path for existing applications:

**The PR fulfills these requirements:**

- [ ] When resolving a specific issue, it's referenced in the PR's title (e.g. `fix #xxx[,#xxx]`, where "xxx" is the issue number)

You have tested in the following browsers: (Providing a detailed version will be better.)

- [ ] Chrome
- [ ] Firefox
- [ ] Safari
- [ ] Edge
- [ ] IE

If adding a **new feature**, the PR's description includes:

- [ ] A convincing reason for adding this feature
- [ ] Related documents have been updated
- [ ] Related tests have been updated

To avoid wasting your time, it's best to open a **feature request issue** first and wait for approval before working on it.

**Other information:**

### CODE REVIEW

**Comment by krmax44** [comment]
There's one thing still missing: the file copying works well, but the config does not yet get altered. To add this, the script needs to know where the layout, components etc. files are located, which is not the same for every theme. I came up with a couple of potential solutions:

- Every theme needs to declare `main` (or some other custom key) in its package.json, which points to a Saber configuration, related: #160 
- The script or theme loader should look for the path containing the `layouts` subdirectory (or a Saber configuration, whatever is a clear sign that this is the theme's root folder). Might be problematic in cases where a dist and source folder both containing a layout folder or other more exotic folder structures.
- Enforce a strict folder structure every theme needs to use
- Check for folders like `dist` and `src`, partially already present but only for npm themes:

https://github.com/saberland/saber/blob/9caaf74eacaac1c9548ebd08170034d33db1c22f/packages/saber/lib/index.js#L140-L144

What do you think? @saberland/core

**Comment by MarchWorks** [comment]
I prefer enforcing a strict folder structure and also adding support for `main` in special cases (I am not sure if there is such special cases but being safe is better)

**Comment by krmax44** [comment]
@egoist your opinion on this?

**Comment by egoist** [comment]
> the file copying works well, but the config does not yet get altered

what config?

**Comment by krmax44** [comment]
@egoist `saber-config.yml`

File: packages/saber/lib/cli.js [hunk 1]
```diff
@@ -54,6 +60,101 @@ cli
...
+
+        log.info('Copied theme from node_modules.')
+      } catch (error) {
+        log.error(`Could not copy files: ${error.toString()}`)
+      }
+    }
+
+    log.info(
+      `Plese change "theme" in your Saber config to "./${relativeDest}".`
+    )
```
Comment (egoist) [inline packages/saber/lib/cli.js]: We should probably remove the old theme from `package.json`?

And reinstall dependencies in the ejected directory?

File: packages/saber/lib/cli.js [hunk 2]
```diff
@@ -54,6 +60,101 @@ cli
...
+      const themePackage = configLoader.load({
+        cwd: themePath,
+        files: ['package.json']
+      }).data
+
+      if (themePackage.repository && themePackage.repository.url) {
+        const repo = themePackage.repository.url
+        await new Promise(resolve =>
+          exec(
+            `git submodule add "${repo}" "${relativeDest}"`,
```
Comment (egoist) [inline packages/saber/lib/cli.js]: is `git submodule add` the proper choice in this case? someone might need to merge the modified theme with upstream?

**Comment by egoist** [comment]
we should also consider monorepo 😄 https://github.com/saberland/create-portfolio/tree/master/packages/saber-theme-portfolio

**Comment by krmax44** [comment]
IIRC, `package.json` allows specifying something like `repository.pathTo`, which we could use for this.

**Comment by krmax44** [comment]
Added monorepo support and dependency merging. Only thing missing is that `saber-config.yml` doesn't get updated accordingly, reasons above. Will need to wait for #160

**Comment by krmax44** [comment]
Okay, I tested it now with portfolio and tailsaw, and it works great so far! What do you think @egoist?

**Pull request merged**